### PR TITLE
Fix CartoDataFrame docs

### PR DESCRIPTION
**Pull request by elenatorro**
Related issue: https://github.com/CartoDB/cartoframes/issues/1214

### CODE REVIEW

File: cartoframes/core/cartodataframe.py [hunk 1]
```diff
@@ -4,17 +4,39 @@
...
 
 class CartoDataFrame(GeoDataFrame):
-
     def __init__(self, *args, **kwargs):
+        """
+        A CartoDataFrame object is a
+        `geopandas.GeoDataFrame <http://geopandas.org/reference/geopandas.GeoDataFrame.html>`_
+        that has a column with geometry. It extends the GeoDataFrame object to read and
+        write data from CARTO, adding wrappers when necessary.
+        """
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: I'm not sure if the constructor `docstring` is being generated, could we double check? In other classes I've added the docs of the constructor just below the class definition.

Comment (alrocar) [reply cartoframes/core/cartodataframe.py]: Adding another comment, but it applies to the rest of the methods of the class. Should we document the *args and **kwargs arguments?

I don't know if it's a good practice or not, but in terms of documentation I think we should be the more explicit possible with the arguments. We might even consider to drop *args and **kwargs in favour of the actual arguments.

Comment (alrocar) [reply cartoframes/core/cartodataframe.py]: Let's add also for each method these sections:

- Args
- Returns
- Raises (let's try to dig a little bit in the code and expose there the most commons exceptions you can get)

File: cartoframes/core/cartodataframe.py [hunk 2]
```diff
@@ -24,16 +46,45 @@ def from_features(cls, features, **kwargs):
...
+
         from ..io.carto import to_carto
         return to_carto(self, *args, **kwargs)
 
     def convert(self, index_column=None, geom_column=None, lnglat_columns=None,
                 drop_index=True, drop_geom=True, drop_lnglat=True):
-        # Magic function
+        """
+        Tries to decode the geometry automatically by looking for coordinates in columns.
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: In this case we can safely add docs for each argument.

File: cartoframes/core/cartodataframe.py [hunk 3]
```diff
@@ -24,16 +46,45 @@ def from_features(cls, features, **kwargs):
...
+
+        Examples:
+
+            Decode the geometry automatically:
+
+            .. code::
+                cdf = CartoDataFrame(data).convert()
+
+            Passing the geometry column explititely:
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]:
```suggestion
            Passing the geometry column explicitly:
```

File: cartoframes/core/cartodataframe.py [hunk 4]
```diff
@@ -24,16 +46,45 @@ def from_features(cls, features, **kwargs):
...
 
     def to_carto(self, *args, **kwargs):
+        """
+        Upload a CartoDataFrame to CARTO.
+        Equivalent to :py:meth:`to_carto <cartoframes.io.to_carto>`.
+
+        Examples:
+
+            .. code::
+                cdf.to_carto(if_exists='replace')
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: Let's include in the code examples some more context, something like this could work:

```python
from cartoframes import CartoDataFrame

cdf = CartoDataFrame.from_file('nybb.shp')
cdf.to_carto(if_exists='replace')
```

File: cartoframes/core/cartodataframe.py [hunk 5]
```diff
@@ -24,16 +46,45 @@ def from_features(cls, features, **kwargs):
...
-        # Magic function
+        """
+        Tries to decode the geometry automatically by looking for coordinates in columns.
+
+        Examples:
+
+            Decode the geometry automatically:
+
+            .. code::
+                cdf = CartoDataFrame(data).convert()
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: - Let's include the import
- What's data?

File: cartoframes/core/cartodataframe.py [hunk 6]
```diff
@@ -24,16 +46,45 @@ def from_features(cls, features, **kwargs):
...
+
+            Passing the geometry column explititely:
+
+            .. code::
+                cdf = CartoDataFrame(data)
+                cdf.convert(geom_column='my_geom_column')
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: - Let's include the import
- What's data?

File: cartoframes/core/cartodataframe.py [hunk 7]
```diff
@@ -24,16 +46,45 @@ def from_features(cls, features, **kwargs):
...
+        """
+
         generate_index(self, index_column, drop_index)
         generate_geometry(self, geom_column, lnglat_columns, drop_geom, drop_lnglat)
         return self
 
     def viz(self, *args, **kwargs):
+        """
+        Create a :py:class:`Map <cartoframes.viz.Map>`. visualization
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: Sometimes we use third singular person and other times plural. (I vote for singular :) )

File: cartoframes/core/cartodataframe.py (hunk 2 above)
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: Let's add some mention that it's decoded as a `shapely` object.

File: cartoframes/core/cartodataframe.py [hunk 8]
```diff
@@ -24,16 +46,45 @@ def from_features(cls, features, **kwargs):
...
         from ..io.carto import to_carto
         return to_carto(self, *args, **kwargs)
 
     def convert(self, index_column=None, geom_column=None, lnglat_columns=None,
                 drop_index=True, drop_geom=True, drop_lnglat=True):
-        # Magic function
+        """
+        Tries to decode the geometry automatically by looking for coordinates in columns.
+
+        Examples:
```
Comment (alrocar) [inline cartoframes/core/cartodataframe.py]: An example with `lnglat_columns` can be quite handy. It seems to me that in many cases people working with external files (specially CSVs) will have those coordinates as separate columns.

File: cartoframes/data/services/geocoding.py [hunk 9]
```diff
@@ -3,375 +3,67 @@
...
-            gc = Geocoding()
-            df = gc.geocode(df, street='address', city='city', country={'value': 'Spain'}, status=['relevance']).data
-            # show rows with relevance greater than 0.7:
-            print(df[df['carto_geocode_relevance']>0.7, axis=1)])
+        dataframe = pandas.read_csv('my_data')
+        dataframe = Geocoding().geocode(dataframe, 'address', cached='my_data').data
 
+    If you execute the previous code multiple times it will only spend credits on the first geocoding;
+    later ones will reuse the results stored in the ``my_data`` table. This will require extra processing
+    time. If the csv file should ever change, cached results will only be applied to unmodified
```
Comment (alrocar) [inline cartoframes/data/services/geocoding.py]:
```suggestion
    time. If the CSV file should ever change, cached results will only be applied to unmodified
```

File: cartoframes/data/services/geocoding.py [hunk 10]
```diff
@@ -401,19 +93,128 @@ def geocode(self, source, street,
...
+            | Name        | Type   | Description                                                |
+            +=============+========+============================================================+
+            | precision   | text   | precise or interpolated                                    |
+            +-------------+--------+------------------------------------------------------------+
+            | relevance   | number | 0 to 1, higher being more relevant                         |
+            +-------------+--------+------------------------------------------------------------+
+            | match_types | array  | list of match type strings                                 |
+            |             |        | point_of_interest, country, state, county, locality,       |
+            |             |        | district, street, intersection, street_number, postal_code |
+            +-------------+--------+------------------------------------------------------------+
```
Comment (alrocar) [inline cartoframes/data/services/geocoding.py]: ❤️

File: cartoframes/data/services/geocoding.py [hunk 11]
```diff
@@ -401,19 +93,128 @@ def geocode(self, source, street,
...
+                geocoded_cdf, metadata = gc.geocode(
+                    df,
+                    street='address',
+                    city='city',
+                    country={'value': 'Spain'},
+                    status=['relevance']
+                )
+
+                # show rows with relevance greater than 0.7:
+                print(geocoded_cdf[geocoded_cdf['carto_geocode_relevance'] > 0.7, axis=1)])
```
Comment (alrocar) [inline cartoframes/data/services/geocoding.py]: ❤️ !

File: examples/data_services/geocoding_dataframe.ipynb [hunk 12]
```diff
@@ -194,9 +177,9 @@
        "  <link href=&quot;https://fonts.googleapis.com/css?family=Roboto&quot; rel=&quot;stylesheet&quot; type=&quot;text/css&quot;>\n",
        "\n",
        "  <!-- External libraries -->\n",
-       "  \n",
-       "  <!-- base64-js -->\n",
-       "  <script src='https://libs.cartocdn.com/cartoframes/dependencies/base64.js'></script>\n",
+       "\n",
+       "  <!-- pako -->\n",
```
Comment (alrocar) [inline examples/data_services/geocoding_dataframe.ipynb]: https://www.youtube.com/watch?v=QZaqZIbJiKo

Comment (elenatorro) [reply examples/data_services/geocoding_dataframe.ipynb]: TBH I don't know where it comes from 😂

File: cartoframes/core/cartodataframe.py (hunk 1 above)
Comment (elenatorro) [reply cartoframes/core/cartodataframe.py]: I haven't added anything related with the *args and **kwargs, I think we should tackle this in a different task 👍

**Comment by elenatorro** [comment]
Updated! Although I've documented the `convert` method, it's usage is still internal, so I've excluded it from the docstring documentation.

File: cartoframes/core/cartodataframe.py [hunk 13]
```diff
@@ -4,36 +4,169 @@
 
 
 class CartoDataFrame(GeoDataFrame):
-
     def __init__(self, *args, **kwargs):
+        """
+        A CartoDataFrame object is a
```
Comment (Jesus89) [inline cartoframes/core/cartodataframe.py]: I would put something like this: (class is more pythonic than object)
```
The CartoDataFrame class is an extension of the `geopandas.GeoDataFrame <http://geopandas.org/reference/geopandas.GeoDataFrame.html>`_ class. It provides powerful cartographic visualizations, geometry detection and decoding, and read / write access to the CARTO platform.
```

File: cartoframes/core/cartodataframe.py [hunk 14]
```diff
@@ -4,36 +4,169 @@
...
+        Upload a CartoDataFrame to CARTO.
+        Equivalent to :py:meth:`to_carto <cartoframes.io.to_carto>`.
+
+        Examples:
+
+            .. code::
+                from cartoframes import CartoDataFrame
+
+                cdf = CartoDataFrame.from_file('nybb.shp')
+                cdf.to_carto(if_exists='replace')
```
Comment (Jesus89) [inline cartoframes/core/cartodataframe.py]: You need to specify here the name of the table: `to_carto('table', if_exists='replace')`.

NOTE: both `from_carto` and `to_carto` methods require a Credentials instance. I would add that in the docstring, and maybe the `set_default_credentials` case too.

File: cartoframes/core/cartodataframe.py [hunk 15]
```diff
@@ -4,36 +4,169 @@
...
+
+        """
+
         generate_index(self, index_column, drop_index)
         generate_geometry(self, geom_column, lnglat_columns, drop_geom, drop_lnglat)
         return self
 
     def viz(self, *args, **kwargs):
+        """
+        Creates a :py:class:`Map <cartoframes.viz.Map>`. visualization
```
Comment (Jesus89) [inline cartoframes/core/cartodataframe.py]: I think there is an extra dot here. I would comment that you can use the Layer API here.

File: cartoframes/data/services/utils/__init__.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+from . import geocoding_constants
+from . import geocoding_utils
+from .table_geocoding_lock import TableGeocodingLock
+
+__all__ = [
```
Comment (Jesus89) [inline cartoframes/data/services/utils/__init__.py]: :heart:

File: cartoframes/io/carto.py [hunk 17]
```diff
@@ -6,6 +6,17 @@
 from ..core.managers.context_manager import ContextManager
 from ..utils.utils import is_sql_query
 
+__all__ = [
```
Comment (Jesus89) [inline cartoframes/io/carto.py]: Why is this added?

File: tests/e2e/data/services/test_isolines.py [hunk 18]
```diff
@@ -116,18 +117,18 @@ def test_isochrones_from_dataframe_dataset(self):
         iso = Isolines(credentials=self.credentials)
 
         df = pd.DataFrame(self.points, columns=['name', 'the_geom'])
-        ds = Dataset(df, credentials=self.credentials)
+        cdf = CartoDataFrame(df, credentials=self.credentials)
```
Comment (Jesus89) [inline tests/e2e/data/services/test_isolines.py]: CartoDataFrame does not have a `credentials` param. Also, you can unify these two lines with the following, because a CDF is also a DF.

```
cdf = CartoDataFrame(self.points, columns=['name', 'the_geom'])
```

File: cartoframes/io/carto.py (hunk 17 above)
Comment (elenatorro) [reply cartoframes/io/carto.py]: To import the methods directly, I couldn't without it, but maybe there's another way.

File: cartoframes/core/cartodataframe.py (hunk 15 above)
Comment (elenatorro) [reply cartoframes/core/cartodataframe.py]: I wouldn't add anything else about the Layer API yet until we consolidate this, it's more an extra for now

File: cartoframes/io/carto.py (hunk 17 above)
Comment (Jesus89) [reply cartoframes/io/carto.py]: Inside the code, we use the full path (cartoframes.io.carto.read_carto). In the public API all these methods are exposed in the top level (cartoframes.read_carto). By default, it should load all these methods without the \_\_all\_\_ variable.

File: cartoframes/core/cartodataframe.py (hunk 15 above)
Comment (Jesus89) [reply cartoframes/core/cartodataframe.py]: Yes, you're right

File: cartoframes/io/carto.py (hunk 17 above)
Comment (elenatorro) [reply cartoframes/io/carto.py]: Oh, I didn't know it! ok, fixed 👍

File: tests/e2e/data/services/test_isolines.py [hunk 19]
```diff
@@ -223,7 +221,8 @@ def test_isochrones_from_dataframe_as_new_table(self):
         self.assertTrue('data_range' in result)
         self.assertEqual(len(result.index), 6)
 
-        cdf = CartoDataFrame.from_carto(table_name, credentials=self.credentials)
+        cdf = CartoDataFrame.from_carto(table_name)
```
Comment (Jesus89) [inline tests/e2e/data/services/test_isolines.py]: Sorry, the Credentials instance is not required in the constructor, but it's required in the `from_carto` and `to_carto` methods.
